drivers/dfplayer/dfplayer_internal.c

```diff
@@ -0,0 +1,488 @@
+             * need to appear at the right position *and* the FCS need to match
+             */
+            if ((data == DFPLAYER_END) && (dev->len == DFPLAYER_LEN)) {
+                uint16_t fcs_exp = fcs_init;
+                fcs_exp -= dev->buf[0] + dev->buf[1] + dev->buf[2] + dev->buf[3];
+                uint16_t fcs = (((uint16_t)dev->buf[4]) << 8) | dev->buf[5];
+                if (fcs == fcs_exp) {
+                    DEBUG("[dfplayer] Got 0x%02x, 0x%02x, 0x%02x, 0x%02x\n",
+                          dev->buf[0], dev->buf[1], dev->buf[2],
+                          dev->buf[3]);
+                    _parse_frame(dev);
+                }
+                else {
+                    DEBUG("[dfplayer] Checksum mismatch");
+                }
+            }
+            else if (dev->len < sizeof(dev->buf)) {
+                dev->buf[dev->len++] = data;
+                return;
+            }
+            else {
+                DEBUG("[dfplayer] Frame overflown\n");
+            }
+            break;
+        default:
+            break;
+    }
+
+    dev->state = DFPLAYER_RX_STATE_START;
+}
+static void _timeout_cb(void *_dev)
+{
+    dfplayer_t *dev = _dev;
+    dev->flags |= DFPLAYER_FLAG_TIMEDOUT;
+    thread_wakeup(dev->sync);
+}
+
+static int _send(dfplayer_t *dev, uint8_t cmd, uint8_t p1, uint8_t p2,
+                  uint32_t timeout_us)
+{
+    int retval;
+    if (dev->flags & DFPLAYER_FLAG_NO_ACK_BUG) {
+        /* Hardware bug: The next command will not be ack'ed, unless it is
+         * a query command. We can clear the flag, as we issue now a fake query,
+         * if needed.
+         */
+        dev->flags &= ~(DFPLAYER_FLAG_NO_ACK_BUG);
+        if (cmd < DFPLAYER_LOWEST_QUERY) {
+            /* Command is a control command, we query the volume and ignore the
+             * result as work around */
+            retval = _send(dev, DFPLAYER_CMD_GET_VOLUME, 0, 0,
+                           DFPLAYER_TIMEOUT_MS * US_PER_MS);
+            if (retval) {
+                /* pass through error */
+                return retval;
+            }
+        }
+    }
+
+    uint16_t fcs = fcs_init - (cmd + DFPLAYER_ACK + p1 + p2);
+    uint8_t frame[] = {
+        DFPLAYER_START, DFPLAYER_VERSION, DFPLAYER_LEN, cmd, DFPLAYER_ACK,
+        p1, p2, (uint8_t)(fcs>>8), (uint8_t)fcs, DFPLAYER_END
+    };
+    xtimer_t timer = {
+        .callback = _timeout_cb,
+        .arg = dev
+    };
+
+    for (unsigned i = 0; i < DFPLAYER_RETRIES; i++) {
+        retval = 0;
+        /* clear timeout flag */
+        dev->flags &= ~(DFPLAYER_FLAG_TIMEDOUT);
+
+        DEBUG("[dfplayer] About to exchange frame\n");
+        /* set timeout timer */
+        xtimer_set(&timer, timeout_us);
+
+        uart_write(dev->uart, frame, sizeof(frame));
+
+        /* wait for reply from ISR */
+        dev->sync = thread_getpid();
+        thread_sleep();
+        dev->sync = KERNEL_PID_UNDEF;
```

Comment:
That's an interesting way to do that.
I would have just called `mutex_lock()` (`xtimer_mutex_lock_timeout()`) here and unlocked it from the interrupt.